File: dairytrack_mobile/lib/controller/APIURL1/cattleDistributionController.dart
import 'dart:convert';
import 'package:dairytrack_mobile/controller/APIURL1/cowManagementController.dart';
import 'package:http/http.dart' as http;
import '../../api/apiController.dart';

class CattleDistributionController {
  final String baseUrl = '$API_URL1/user-cow'; // Adjust base URL as needed
  final Map<String, String> _headers = {'Content-Type': 'application/json'};

  // Assign a cow to a user
  Future<Map<String, dynamic>> assignCowToUser(int userId, int cowId) async {
    try {
      final url = Uri.parse('$baseUrl/assign');
      final response = await http.post(
        url,
        headers: _headers,
        body: jsonEncode({'user_id': userId, 'cow_id': cowId}),
      );

      return _handleResponse(response, 'Failed to assign cow to user');
    } catch (e) {
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  // Unassign a cow from a user
  Future<Map<String, dynamic>> unassignCowFromUser(
      int userId, int cowId) async {
    try {
      final url = Uri.parse('$baseUrl/unassign');
      final response = await http.post(
        url,
        headers: _headers,
        body: jsonEncode({'user_id': userId, 'cow_id': cowId}),
      );

      return _handleResponse(response, 'Failed to unassign cow from user');
    } catch (e) {
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  // List cows managed by a user
  Future<Map<String, dynamic>> listCowsByUser(int userId) async {
    try {
      final url = Uri.parse('$baseUrl/list/$userId');
      final response = await http.get(url, headers: _headers);

      print('API URL: $url'); // Debug print
      print('Response status: ${response.statusCode}'); // Debug print
      print('Response body: ${response.body}'); // Debug print

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body) as Map<String, dynamic>;

        // Convert cow data to proper format
        List<Map<String, dynamic>> cows = [];
        if (responseData['cows'] != null) {
          cows = (responseData['cows'] as List)
              .map((cowData) => Map<String, dynamic>.from(cowData as Map))
              .toList();
        } else if (responseData['data'] != null &&
            responseData['data']['cows'] != null) {
          cows = (responseData['data']['cows'] as List)
              .map((cowData) => Map<String, dynamic>.from(cowData as Map))
              .toList();
        }

        return {'success': true, 'cows': cows, 'data': responseData};
      } else {
        return {'success': false, 'message': 'Failed to list cows by user'};
      }
    } catch (e) {
      print('Error in listCowsByUser: $e');
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  // Get farmers with cows
  Future<Map<String, dynamic>> getFarmersWithCows() async {
    try {
      final url = Uri.parse('$baseUrl/farmers-with-cows');
      final response = await http.get(url, headers: _headers);

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body) as Map<String, dynamic>;
        return {'success': true, 'data': responseData};
      } else {
        return {'success': false, 'message': 'Failed to get farmers with cows'};
      }
    } catch (e) {
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  // Get users with cows - formatted for analysis view
  Future<Map<String, dynamic>> getUsersWithCows() async {
    try {
      final url = Uri.parse('$baseUrl/farmers-with-cows');
      final response = await http.get(url, headers: _headers);

      print('API URL: $url');
      print('Response status: ${response.statusCode}');
      print('Response body: ${response.body}');

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body) as Map<String, dynamic>;

        // Transform data to match the expected format for analysis view
        List<Map<String, dynamic>> usersWithCows = [];

        if (responseData['farmers_with_cows'] != null) {
          usersWithCows =
              (responseData['farmers_with_cows'] as List).map((farmer) {
            final farmerMap = Map<String, dynamic>.from(farmer as Map);

            // Transform user data
            Map<String, dynamic> userData = {};
            if (farmerMap['user'] != null) {
              userData = Map<String, dynamic>.from(farmerMap['user'] as Map);
            }

            // Transform cows data and add farmer information
            List<Map<String, dynamic>> cows = [];
            if (farmerMap['cows'] != null) {
              cows = (farmerMap['cows'] as List).map((cow) {
                final cowMap = Map<String, dynamic>.from(cow as Map);
                return {
                  ...cowMap,
                  'farmerName':
                      userData['name'] ?? userData['username'] ?? 'Unknown',
                  'farmerId': userData['id'],
                };
              }).toList();
            }

            return {
              'id': userData['id'],
              'name': userData['name'] ?? userData['username'],
              'username': userData['username'],
              'email': userData['email'],
              'contact': userData['contact'],
              'religion': userData['religion'],
              'role_id': userData['role_id'],
              'token': userData['token'],
              'cows': cows,
            };
          }).toList();
        }

        return {
          'success': true,
          'usersWithCows': usersWithCows,
          'data': responseData
        };
      } else {
        return {'success': false, 'message': 'Failed to get users with cows'};
      }
    } catch (e) {
      print('Error in getUsersWithCows: $e');
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  // Get all users and all cows
  Future<Map<String, dynamic>> getAllUsersAndAllCows() async {
    try {
      final url = Uri.parse('$baseUrl/all-users-and-all-cows');
      final response = await http.get(url, headers: _headers);

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body) as Map<String, dynamic>;
        return {'success': true, 'data': responseData};
      } else {
        return {
          'success': false,
          'message': 'Failed to get all users and all cows'
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  Future<Map<String, dynamic>> getCowManagers(int cowId) async {
    try {
      final url = Uri.parse('$baseUrl/cow-managers/$cowId');
      final response = await http.get(url, headers: _headers);

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body) as Map<String, dynamic>;
        return {'success': true, 'managers': responseData['managers']};
      } else {
        return {'success': false, 'message': 'Failed to get cow managers'};
      }
    } catch (e) {
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  // Get farmers (managers) for a specific cow
  Future<Map<String, dynamic>> getFarmersForCow(int cowId) async {
    try {
      final url = Uri.parse('$baseUrl/cow-managers/$cowId');
      final response = await http.get(url, headers: _headers);

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body) as Map<String, dynamic>;
        return {
          'success': true,
          'cow_id': responseData['cow_id'],
          'farmers': responseData['managers']
        };
      } else {
        return {'success': false, 'message': 'Failed to get farmers for cow'};
      }
    } catch (e) {
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  // Generic response handler
  Map<String, dynamic> _handleResponse(
      http.Response response, String errorMessage) {
    if (response.statusCode >= 200 && response.statusCode < 300) {
      final responseData = jsonDecode(response.body) as Map<String, dynamic>;
      return {'success': true, 'data': responseData};
    } else {
      try {
        final errorData = jsonDecode(response.body) as Map<String, dynamic>;
        return {
          'success': false,
          'message': errorData['error'] ?? errorMessage,
        };
      } catch (e) {
        return {'success': false, 'message': errorMessage};
      }
    }
  }
}

File: dairyTrack-web/src/Modules/controllers/cattleDistributionController.js
import { API_URL1 } from "../../api/apiController.js";
import Swal from "sweetalert2"; // Import SweetAlert

// Fungsi untuk menambahkan relasi antara User dan Cow (dengan alert)
export const assignCowToUser = async (userId, cowId) => {
  const result = await assignCowToUserSilent(userId, cowId);

  if (result.success) {
    Swal.fire({
      icon: "success",
      title: "Success",
      text: result.message,
    });
  } else {
    Swal.fire({
      icon: "error",
      title: "Error",
      text: result.message,
    });
  }

  return result;
};

// Fungsi untuk menambahkan relasi antara User dan Cow (tanpa alert - untuk bulk operations)
export const assignCowToUserSilent = async (userId, cowId) => {
  try {
    const response = await fetch(`${API_URL1}/user-cow/assign`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ user_id: userId, cow_id: cowId }),
    });

    if (response.ok) {
      const data = await response.json();
      return { success: true, message: data.message };
    } else if (response.status === 500) {
      return {
        success: false,
        message: "This cow is already assigned to this user.",
      };
    } else {
      const error = await response.json();
      return {
        success: false,
        message: error.error || "Failed to assign cow to user.",
      };
    }
  } catch (error) {
    console.error("Error assigning cow to user:", error);
    return {
      success: false,
      message: "An error occurred while assigning the cow to the user.",
    };
  }
};

// Fungsi untuk menghapus relasi antara User dan Cow (dengan alert)
export const unassignCowFromUser = async (userId, cowId) => {
  const result = await unassignCowFromUserSilent(userId, cowId);

  if (result.success) {
    Swal.fire({
      icon: "success",
      title: "Success",
      text: result.message,
    });
  } else {
    Swal.fire({
      icon: "error",
      title: "Error",
      text: result.message,
    });
  }

  return result;
};

// Fungsi untuk menghapus relasi antara User dan Cow (tanpa alert - untuk bulk operations)
export const unassignCowFromUserSilent = async (userId, cowId) => {
  try {
    const response = await fetch(`${API_URL1}/user-cow/unassign`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ user_id: userId, cow_id: cowId }),
    });

    if (response.ok) {
      const data = await response.json();
      return { success: true, message: data.message };
    } else {
      const error = await response.json();
      return {
        success: false,
        message: error.error || "Failed to unassign cow from user.",
      };
    }
  } catch (error) {
    console.error("Error unassigning cow from user:", error);
    return {
      success: false,
      message: "An error occurred while unassigning the cow from the user.",
    };
  }
};

// Fungsi untuk mendapatkan daftar sapi yang dikelola oleh User tertentu
export const listCowsByUser = async (userId) => {
  try {
    const response = await fetch(`${API_URL1}/user-cow/list/${userId}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (response.ok) {
      const data = await response.json();
      return { success: true, cows: data.cows }; // Kembalikan daftar sapi
    } else {
      const error = await response.json();
      return {
        success: false,
        message: error.error || "Failed to fetch cows.",
      };
    }
  } catch (error) {
    Swal.fire({
      icon: "error",
      title: "Error",
      text: "An error occurred while fetching the cows.",
    });
    console.error("Error fetching cows:", error);
    return {
      success: false,
      message: "An error occurred while fetching the cows.",
    };
  }
};

export const getUsersWithCows = async () => {
  try {
    const response = await fetch(`${API_URL1}/user-cow/farmers-with-cows`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (response.ok) {
      const data = await response.json();
      return { success: true, usersWithCows: data.farmers_with_cows }; // Sesuaikan dengan nama properti API
    } else {
      const error = await response.json();
      return {
        success: false,
        message: error.error || "Failed to fetch users with cows.",
      };
    }
  } catch (error) {
    console.error("Error fetching users with cows:", error);
    return {
      success: false,
      message: "An error occurred while fetching users with cows.",
    };
  }
};
// Fungsi untuk mendapatkan semua pengguna dan semua sapi
export const getAllUsersAndAllCows = async () => {
  try {
    const response = await fetch(
      `${API_URL1}/user-cow/all-users-and-all-cows`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      }
    );

    if (response.ok) {
      const data = await response.json();
      // Filter users with role_id === 3
      const filteredUsers = data.users.filter((user) => user.role_id === 3);
      return { success: true, users: filteredUsers, cows: data.cows };
    } else {
      const error = await response.json();
      return {
        success: false,
        message: error.error || "Failed to fetch all users and all cows.",
      };
    }
  } catch (error) {
    console.error("Error fetching all users and all cows:", error);
    return {
      success: false,
      message: "An error occurred while fetching all users and all cows.",
    };
  }
};

export const getCowManagers = async (cowId) => {
  try {
    const response = await fetch(`${API_URL1}/user-cow/cow-managers/${cowId}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (response.ok) {
      const data = await response.json();
      return { success: true, managers: data.managers }; // Kembalikan daftar manager
    } else {
      const error = await response.json();
      return {
        success: false,
        message: error.error || "Failed to fetch cow managers.",
      };
    }
  } catch (error) {
    console.error("Error fetching cow managers:", error);
    return {
      success: false,
      message: "An error occurred while fetching cow managers.",
    };
  }
};

// Fungsi untuk multiple assign dengan summary alert
export const assignMultipleCowsToUser = async (userId, cowIds) => {
  const results = [];
  let successCount = 0;
  let errorCount = 0;
  const errors = [];

  for (const cowId of cowIds) {
    const result = await assignCowToUserSilent(userId, cowId);
    results.push({ cowId, ...result });

    if (result.success) {
      successCount++;
    } else {
      errorCount++;
      errors.push(`Cow ID ${cowId}: ${result.message}`);
    }
  }

  // Show summary alert
  if (errorCount === 0) {
    Swal.fire({
      icon: "success",
      title: "All Assignments Successful",
      text: `Successfully assigned ${successCount} cow(s) to user.`,
    });
  } else if (successCount === 0) {
    Swal.fire({
      icon: "error",
      title: "All Assignments Failed",
      html: `Failed to assign all ${errorCount} cow(s):<br><br>${errors.join(
        "<br>"
      )}`,
    });
  } else {
    Swal.fire({
      icon: "warning",
      title: "Partial Success",
      html: `Successfully assigned: ${successCount} cow(s)<br>Failed: ${errorCount} cow(s)<br><br>Errors:<br>${errors.join(
        "<br>"
      )}`,
    });
  }

  return {
    success: errorCount === 0,
    successCount,
    errorCount,
    results,
    errors,
  };
};

// Fungsi untuk multiple unassign dengan summary alert
export const unassignMultipleCowsFromUser = async (userId, cowIds) => {
  const results = [];
  let successCount = 0;
  let errorCount = 0;
  const errors = [];

  for (const cowId of cowIds) {
    const result = await unassignCowFromUserSilent(userId, cowId);
    results.push({ cowId, ...result });

    if (result.success) {
      successCount++;
    } else {
      errorCount++;
      errors.push(`Cow ID ${cowId}: ${result.message}`);
    }
  }

  // Show summary alert
  if (errorCount === 0) {
    Swal.fire({
      icon: "success",
      title: "All Unassignments Successful",
      text: `Successfully unassigned ${successCount} cow(s) from user.`,
    });
  } else if (successCount === 0) {
    Swal.fire({
      icon: "error",
      title: "All Unassignments Failed",
      html: `Failed to unassign all ${errorCount} cow(s):<br><br>${errors.join(
        "<br>"
      )}`,
    });
  } else {
    Swal.fire({
      icon: "warning",
      title: "Partial Success",
      html: `Successfully unassigned: ${successCount} cow(s)<br>Failed: ${errorCount} cow(s)<br><br>Errors:<br>${errors.join(
        "<br>"
      )}`,
    });
  }

  return {
    success: errorCount === 0,
    successCount,
    errorCount,
    results,
    errors,
  };
};

// Fungsi untuk assign satu user ke multiple cows dengan summary alert
export const assignUserToMultipleCows = async (userId, cowIds) => {
  const results = [];
  let successCount = 0;
  let errorCount = 0;
  const errors = [];

  for (const cowId of cowIds) {
    const result = await assignCowToUserSilent(userId, cowId);
    results.push({ cowId, ...result });

    if (result.success) {
      successCount++;
    } else {
      errorCount++;
      errors.push(`Cow ID ${cowId}: ${result.message}`);
    }
  }

  // Show summary alert
  if (errorCount === 0) {
    Swal.fire({
      icon: "success",
      title: "Bulk Assignment Successful",
      text: `User successfully assigned to ${successCount} cow(s).`,
    });
  } else if (successCount === 0) {
    Swal.fire({
      icon: "error",
      title: "Bulk Assignment Failed",
      html: `Failed to assign user to all ${errorCount} cow(s):<br><br>${errors.join(
        "<br>"
      )}`,
    });
  } else {
    Swal.fire({
      icon: "warning",
      title: "Partial Assignment Success",
      html: `Successfully assigned to: ${successCount} cow(s)<br>Failed: ${errorCount} cow(s)<br><br>Errors:<br>${errors.join(
        "<br>"
      )}`,
    });
  }

  return {
    success: errorCount === 0,
    successCount,
    errorCount,
    results,
    errors,
  };
};
